homework_solution.md

```diff
@@ -0,0 +1,474 @@
+
+      ```
+      ALTER TABLE trips_backup
+      DROP COLUMN from_station_name,
+      DROP COLUMN to_station_name;
+      ```
+
+   3. Before dropping the columns, we have to populate the station ids columns where they're null
+
+      ```
+      UPDATE trips t
+      SET from_station_id = (SELECT id FROM stations s WHERE t.from_station_name = s.name),
+      to_station_id  = (SELECT id FROM stations s WHERE t.to_station_name = s.name)
+      WHERE from_station_id IS NULL OR to_station_id IS NULL
+      ```
+
+Now all of our `trips` should have stations ids that can be joined on the `stations` table when we need to get more information.
+
+# Part 2: Missing Date Data
+
+1.  _**What's the inconsistency in date formats? You can assume that each quarter's trips are numbered sequentially, starting with the first day of the first month of that quarter.**_
+
+    There are many. We can examine them by running (this uses the fairly safe assumption of consistency within each of the original csv, so looking at the first record for a given file provides insight into the whole file, and looking into the start date format gives insight into the end date format).
+
+    ```
+    SELECT original_filename, (ARRAY_AGG(DISTINCT start_time_str))[1]
+    FROM trips
+    GROUP BY original_filename;
+    ```
+
+    It gets us the result
+
+    ```
+    original_filename             |     array_agg
+    ------------------------------------------+-------------------
+    Bikeshare Ridership (2017 Q1).csv        | 10/1/2017 0:03
+    Bikeshare Ridership (2017 Q2).csv        | 10/4/2017 0:00
+    Bikeshare Ridership (2017 Q3).csv        | 7/10/2017 0:00
+    Bikeshare Ridership (2017 Q4).csv        | 10/01/17 00:00:01
+    Bike Share Toronto Ridership_Q1 2018.csv | 1/10/2018 0:01
+    Bike Share Toronto Ridership_Q2 2018.csv | 4/10/2018 0:01
+    Bike Share Toronto Ridership_Q3 2018.csv | 7/10/2018 0:00
+    Bike Share Toronto Ridership_Q4 2018.csv | 10/10/2018 0:01
+    ```
+
+    We can see there that:
+
+    - Throughout 2017 there's no consistency in using `dd/mm` and `mm/dd`. The first 2 quarters use `dd/m`, while the last 2 use `m/dd`.
+    - Further, throughout 2017 there's no consistency in using `yy` and `yyyy`. The first 3 quarters use the full year, while the last one uses short year.
+    - Time format is also inconsistent throughout 2017, with the first 3 quarters using `h:mm`, while the last one uses `hh:mm:ss`
+    - 2018 was a beautiful year for date formating in the bike share organization
+
+2.  _**Take a look at Postgres's [date functions](https://www.postgresql.org/docs/12/functions-datetime.html), and fill in the missing date data using proper timestamps. You may have to write several queries to do this.**_
+
+    Our final goal is to get all dates to be in the format used throughout 2018 files, which is `mm/dd/yyyy h:mm`.
```

Comment:
[question]
Why this format, and not, say,[ ISO-8601](https://en.wikipedia.org/wiki/ISO_8601)?

Reply:
The assumption here was that the format stabilized by 2018, and that was the desired format by the organization.